### PR TITLE
Add certificate rotation to host agent 

### LINKED ISSUE #555: Manage certificate rotations
To enhance security, we need a mechanism to rotate our certificates to be updated sometime before their expiry.
These certs will primarily be used by kubeconfig. So we need a process to update these kubeconfigs with new certificates and also the client is to be updated with new kubconfig.

Part of #484
 
Ref: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20210222-kubelet-authentication.md#kubelet-authenticator-flow

### PR DESCRIPTION
**Pull request by sachinkumarsingh092**
**What this PR does / why we need it**:
To enhance security, we need a mechanism to rotate our certificates to be updated sometime before their expiry. We'll employ the bootstrap config creation workflow to create new certificates as part of the new kubeconfig. 
We introduce a new flag `certExpiryDuration` in the agent to be used to pass the certificate expiration time from the current creation time.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #555

**Additional information**

File changes:
- **agent/help_flag_test.go**: Minor changes to fix tests.
- **agent/main.go**: The certification rotation logic checks the time left by the current certificate and initiates the bootstrap flow for new config creation if less than 20% time remains for expiry.
- **agent/registration/csr.go**: New argument to the `NewByohCSR()` function and its respective calls.
- **agent/registration/csr_test.go**: Test fixes for the above.
- **hack/install-host-agent-service.sh**: This script is used to start the agent as a systemd service that restarts the agent whenever the kubeconfig changes. This is done to mitigate the problem of terminating the process during the change in kubeconfig which leads to restarting the manager and killing the process. This script allows the host agent service to be restarted after process termination, and a watcher service observes the kubeconfig for changes.

**Special notes for your reviewer**
We cannot write integration tests for this PR due to the large minimum required time for certificate expiration, which is 10 minutes. So we will have a separate PR for those slow e2e tests. For this reason, we have placed a feature flag around the cert rotation section for now.

### CODE REVIEW

File: agent/main.go [hunk 1]
```diff
@@ -231,3 +242,53 @@ func handleBootstrapFlow(logger logr.Logger, hostName string) error {
...
+
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			logger.Error(err, "Certifcate parse failed")
+			return err
+		}
+
+		totalTimeCert := cert.NotAfter.Sub(cert.NotBefore)
+
+		// if less than 20% time left, renew the certs
```
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply agent/main.go]: Done

File: agent/main.go [hunk 2]
```diff
@@ -231,3 +242,53 @@ func handleBootstrapFlow(logger logr.Logger, hostName string) error {
...
+			logger.Info("certificate expiration time left is less than 20%, renewing")
+			if err = handleBootstrapFlow(logger, hostName); err != nil {
+				logger.Error(err, "bootstrap flow failed")
+			}
+		} else {
+			logger.Info("certificate are valid", "will be renewed after", cert.NotAfter.Add(totalTimeCert/-5))
+		}
+
+		// Poll after every 4 seconds
+		time.Sleep(4 * time.Second) //nolint
```
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply agent/main.go]: Done

File: agent/main.go [hunk 3]
```diff
@@ -231,3 +242,53 @@ func handleBootstrapFlow(logger logr.Logger, hostName string) error {
...
+	}
+}
+
+func getConfig(logger logr.Logger) *rest.Config {
```
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply agent/main.go:279]: Yes. It does make sense to break main() into more logical units. When we have enough of these units, we shall move them to a new package.

File: agent/registration/csr_test.go [hunk 4]
```diff
@@ -104,7 +104,7 @@ kovW9X7Ook/tTW0HyX6D6HRciA==
 		})
 
 		It("should return error if hostname is invalid", func() {
-			CSRRegistrar, err := registration.NewByohCSR(cfg, logr.Discard())
+			CSRRegistrar, err := registration.NewByohCSR(cfg, logr.Discard(), int64((time.Hour * 24).Seconds()))
```
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply agent/registration/csr_test.go]: done

File: hack/install-host-agent-service.sh [hunk 5]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply hack/install-host-agent-service.sh:1]: Done. Have a look.

File: agent/main.go [hunk 6]
```diff
@@ -156,23 +163,27 @@ func main() {
...
 	registration.LocalHostRegistrar = &registration.HostRegistrar{K8sClient: k8sClient}
 	err = registration.LocalHostRegistrar.Register(hostName, namespace, labels)
 	if err != nil {
 		logger.Error(err, "error registering host %s registration in namespace %s", hostName, namespace)
 		return
 	}
 
+	// Start certificate rotation goroutine.
+	// This is behind a feature flag for now. Set 'CERTIFICATE_ROTATION=true' to enable it.
+	if os.Getenv("CERTIFICATE_ROTATION") == "true" {
```
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply agent/main.go:177]: We tried to run the agent standalone and store logs for it, similar to how we did for the [--version flag test](https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/blob/main/agent/host_agent_test.go#L346). But running agent without any such informational flag causes it to run indefinitely without any termination. Hence, normal string matching and even the normal defining the flag at os level and checking it won't work. The best way out of it is to write e2e imo.

File: agent/main.go [hunk 7]
```diff
@@ -83,6 +87,7 @@ func setupflags() {
 	klog.ClearLogger()
 
 	flag.StringVar(&namespace, "namespace", "default", "Namespace in the management cluster where you would like to register this host")
+	flag.Int64Var(&certExpiryDuration, "certExpiryDuration", registration.ExpirationSeconds, "Duration for the expiration of the host certificates")
```
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply agent/main.go:90]: This is interrelated to the above comment and suffers a similar problem.

File: agent/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anusha94) [reply agent/main.go:177]: Why would you want it to terminate? We should be able to write an `Eventually` block that will read from the logs and mark the test as successful.

> running agent without any such informational flag

what are informational flags? Build flags? But we want to set an env variable, right?

File: docs/getting_started.md [hunk 8]
```diff
@@ -226,7 +226,12 @@ NAME                                                          STATUS     ROLES
 byoh-cluster-8siai8                                           Ready      master   5m   v1.23.5
 ```
 
+## Additional: Running host-agent as a systemd service
+You can use the script `hack/install-host-agent-service.sh` to start the agent as a systemd service that restarts the agent whenever the kubeconfig changes. This is done to mitigate the problem of terminating the process during the change in kubeconfig which leads to restarting the manager and killing the process. This script allows the host agent service to be restarted after process termination, and a watcher service observes the kubeconfig for changes. This script requires superuser privillages for its execution.
```
Comment (anusha94) [inline docs/getting_started.md:230]: I'd rephrase this to be how and why it would be useful if you start as a systemd service and not in terms of mitigating something

File: agent/main.go (hunk 7 above)
Comment (Madhur97) [inline agent/main.go:90]: Can we also specify the unit of duration in flag description?

What is the expiry value you expect customers would use? Kubernetes certificates generally have an expiry of an year. If we expect the same range, do you think users would prefer to specify expiry duration in days instead of seconds.

File: agent/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anusha94) [reply agent/main.go:177]: In the interest of not stalling this PR further and given that we will write an e2e test, I think we can skip the test for now.
@sachinkumarsingh092 Please open an issue to write e2e and any necessary unit / integration tests

File: docs/getting_started.md (hunk 8 above)
Comment (sachinkumarsingh092) [reply docs/getting_started.md]: Done. Please have a look.

File: agent/main.go (hunk 7 above)
Comment (sachinkumarsingh092) [reply agent/main.go]: Updated the description.

File: agent/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sachinkumarsingh092) [reply agent/main.go:177]: Done. Raised issue #656

**Review by anusha94 — APPROVED** [review]
lgtm

**Comment by Madhur97** [comment]
lgtm

**Review by Madhur97 — APPROVED** [review]

**Review by Madhur97 — APPROVED** [review]
